### DISCUSSION TITLE
How to keep state between app restarts?

### DISCUSSION THREAD

**Discussion by lgfischer** [vladimir-bukhtoyarov/bucket4j#250] [Q&A] [answered]
Hi,

I'm trying to learn how to integrate bucket4j on our API. One requisite is that I must keep the rate-limit state between app restarts. For example, if I define a limit of 1 calls per hour, deploying our app should not restart the rate-limit counters (even though the app itself will restart).

Is there any simple path to fulfill this requisite? Any way to load/save the state from/to a file should work.

Kind regards

**Comment by vladimir-bukhtoyarov** [comment]
Hello @lgfischer 

Do you to persist state only for graceful restart? What would you plan to do when your application crashed unexpectedly(as result bucket state is not persisted between restart)?

**Comment by lgfischer** [comment]
Hi @vladimir-bukhtoyarov, thank you for your help.

I'm looking to save the rate-limit state every minute and before the app restart. We do daily deploys, so graceful app restarts are very common. Our app rarely crashes, and losing the last-minute state changes is acceptable in these cases.

Kind regards

**Reply by vladimir-bukhtoyarov** [reply]
I have performed quick investigation of currently available abilities and did not find correct way to achieve your requirements.

It looks that desired functionality can be achieved by extending the API of [LocalBucket](https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/local/LocalBucket.java)
```java

public byte[] toSnapshot();

static LocalBucket fromSnapshot(byte[] snapshot) {
    ...
}

```

Most likely I will have a time to implement this feature at the beginning of May.

**Reply by lgfischer** [reply]
Thank you very much, @vladimir-bukhtoyarov! I will investigate this API.

Kind regards

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
@lgfischer 
Discussed functionality has been released with version 7.5.0

See:
* https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/local/LocalBucket.java#L61
* https://github.com/vladimir-bukhtoyarov/bucket4j/blob/master/bucket4j-core/src/main/java/io/github/bucket4j/local/LocalBucket.java#L72

**Reply by lgfischer** [reply]
@vladimir-bukhtoyarov thank you very much!
